index.js

```diff
@@ -1,93 +1,138 @@
-      this.message += ': ' + cause.message
-    }
     return this
   }
-}
 
-/**
- * Return the `info` property from `error` and recursively merge the `info`
- * properties from the error's causes, if any.
- *
- * If a property is repeated, the first one in the cause chain wins.
- *
- * @param {?Error} error assumed not to have circular causes
- * @return {Object}
- */
-function getFullInfo(error) {
-  if (!error) return {}
-  const info = getFullInfo(error.cause)
-  if (typeof error.info === 'object') Object.assign(info, error.info)
-  return info
-}
+  /**
+   * Tag debugging information onto any error (whether an OError or not) and
+   * return it.
+   *
+   * @param {Error} error the error to tag
+   * @param {string} [message] message with which to tag `error`
+   * @param {Object} [info] extra data with wich to tag `error`
+   * @return {Error} the modified `error` argument
+   */
+  static tag(error, message, info) {
+    const oError = /** @type{OError} */ (error)
 
-/**
- * Return the `stack` property from `error` and recursively append the `stack`
- * properties from the error's causes, if any.
- *
- * @param {?Error} error assumed not to have circular causes
- * @return {string}
- */
-function getFullStack(error) {
-  if (!error) return ''
-  const causeStack = getFullStack(error.cause)
-  if (causeStack) return error.stack + '\ncaused by: ' + causeStack
-  return error.stack
+    if (!oError._oErrorTags) oError._oErrorTags = []
+
+    let tag
+    if (Error.captureStackTrace) {
+      // Hide this function in the stack trace, and avoid capturing it twice.
+      tag = /** @type TaggedError */ ({ name: 'TaggedError', message, info })
+      Error.captureStackTrace(tag, OError.tag)
+    } else {
+      tag = new TaggedError(message, info)
+    }
+
+    oError._oErrorTags.push(tag)
+
+    return error
+  }
+
+  /**
+   * The merged info from any `tag`s on the given error.
+   *
+   * If an info property is repeated, the last one wins.
+   *
+   * @param {Error | null | undefined} error any errror (may or may not be an `OError`)
+   * @return {Object}
+   */
+  static getFullInfo(error) {
+    const info = {}
+
+    if (!error) return info
+
+    const oError = /** @type{OError} */ (error)
+
+    if (typeof oError.info === 'object') Object.assign(info, oError.info)
+
+    if (oError._oErrorTags) {
+      for (const tag of oError._oErrorTags) {
+        Object.assign(info, tag.info)
+      }
+    }
+
+    return info
+  }
+
+  /**
+   * Return the `stack` property from `error`, including the `stack`s for any
+   * tagged errors added with `OError.tag` and for any `cause`s.
+   *
+   * @param {Error | null | undefined} error any error (may or may not be an `OError`)
+   * @return {string}
+   */
+  static getFullStack(error) {
+    if (!error) return ''
+
+    const oError = /** @type{OError} */ (error)
+
+    let stack = oError.stack
+
+    if (oError._oErrorTags) {
+      for (const tag of oError._oErrorTags) {
```

Comment:
could be one-lined: `oError._oErrorTags.map(tag => tag.stack).join('\n')`

Reply:
Right, done in 38dc7e9.